File: Frontend/src/services/labelService.ts
// labelService.ts
import api, { PageDTO } from './api';
import { Label } from './taskService';

export interface LabelCreateRequest {
  name: string;
  color: string;
  projectId?: string;
  createdById?: string;
}

class LabelService {
  private basePath = '/api/v1/taskfarm/labels';

  /**
   * Get all labels
   */
  public async getAllLabels(page: number = 1): Promise<PageDTO<Label>> {
    return await api.get<PageDTO<Label>>(`${this.basePath}?page=${page}`);
  }

  /**
   * Get labels by project
   */
  public async getLabelsByProject(projectId: string): Promise<Label[]> {
    return await api.get<Label[]>(`${this.basePath}/project/${projectId}`);
  }

  /**
   * Create a new label
   */
  public async createLabel(label: LabelCreateRequest): Promise<Label> {
    return await api.post<Label>(this.basePath, label);
  }

  /**
   * Update an existing label
   */
  public async updateLabel(id: string, label: LabelCreateRequest): Promise<Label> {
    return await api.put<Label>(`${this.basePath}/${id}`, label);
  }

  /**
   * Delete a label
   */
  public async deleteLabel(id: string): Promise<void> {
    await api.delete(`${this.basePath}/${id}`);
  }

  /**
   * Add label to task
   */
  public async addLabelToTask(taskId: string, labelId: string): Promise<void> {
    await api.post(`${this.basePath}/tasks/${taskId}/labels/${labelId}`, {});
  }

  /**
   * Remove label from task
   */
  public async removeLabelFromTask(taskId: string, labelId: string): Promise<void> {
    await api.delete(`${this.basePath}/tasks/${taskId}/labels/${labelId}`);
  }

  /**
   * Get labels for task
   */
  public async getLabelsForTask(taskId: string): Promise<Label[]> {
    return await api.get<Label[]>(`${this.basePath}/tasks/${taskId}`);
  }
}

const labelService = new LabelService();
export default labelService;

File: Backend/src/main/java/ir/netpick/platform/taskfarm/service/LabelService.java
package ir.netpick.platform.taskfarm.service;

import ir.netpick.platform.core.PageDTO;
import ir.netpick.platform.core.constants.GeneralConstants;
import ir.netpick.platform.core.exception.ResourceNotFoundException;
import ir.netpick.platform.core.utils.PageDTOMapper;
import ir.netpick.platform.taskfarm.model.Label;
import ir.netpick.platform.taskfarm.model.Task;
import ir.netpick.platform.taskfarm.model.TaskLabel;
import ir.netpick.platform.taskfarm.repository.LabelRepository;
import ir.netpick.platform.taskfarm.repository.TaskLabelRepository;
import ir.netpick.platform.taskfarm.repository.TaskRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class LabelService {

    private final LabelRepository labelRepository;
    private final TaskLabelRepository taskLabelRepository;
    private final TaskRepository taskRepository;

    public PageDTO<Label> getAll(int pageNumber) {
        Pageable pageable = PageRequest.of(pageNumber - 1, GeneralConstants.PAGE_SIZE,
                Sort.by("createdAt").descending());
        Page<Label> page = labelRepository.findByDeletedFalse(pageable);
        return PageDTOMapper.map(page);
    }

    public List<Label> getByProject(UUID projectId) {
        return labelRepository.findByProjectIdAndDeletedFalse(projectId);
    }

    public Label getById(UUID labelId) {
        return labelRepository.findById(labelId)
                .orElseThrow(() -> new ResourceNotFoundException("Label with id [%s] not found".formatted(labelId)));
    }

    public Label create(Label label) {
        label.setId(null);
        return labelRepository.save(label);
    }

    public Label update(UUID labelId, Label label) {
        Label existing = getById(labelId);
        label.setId(labelId);
        label.setCreatedAt(existing.getCreatedAt());
        return labelRepository.save(label);
    }

    public void delete(UUID labelId) {
        labelRepository.deleteById(labelId);
    }

    public void addLabelToTask(UUID taskId, UUID labelId) {
        Task task = taskRepository.findById(taskId)
                .orElseThrow(() -> new ResourceNotFoundException("Task not found"));
        Label label = getById(labelId);
        TaskLabel taskLabel = new TaskLabel(taskId, labelId);
        taskLabel.setTask(task);
        taskLabel.setLabel(label);
        taskLabelRepository.save(taskLabel);
    }

    public void removeLabelFromTask(UUID taskId, UUID labelId) {
        taskLabelRepository.deleteByTaskIdAndLabelId(taskId, labelId);
    }

    public List<Label> getLabelsForTask(UUID taskId) {
        List<TaskLabel> taskLabels = taskLabelRepository.findByTaskId(taskId);
        return taskLabels.stream()
                .map(TaskLabel::getLabel)
                .toList();
    }
}